File: Python3/496. Next Greater Element I.py
class Solution:
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        l = [-1] * len(nums1)
        for i in range(len(nums1)):
            k = 0
            for k in range(len(nums2)):
                if nums2[k] == nums1[i]:
                    break
            for j in range(k + 1, len(nums2)):
                if nums2[j] > nums1[i]:
                    l[i] = nums2[j]
                    break
        return l


sol = Solution()
print(sol.nextGreaterElement([4, 1, 2], [1, 3, 4, 2]))
print(sol.nextGreaterElement([2, 4], [1, 2, 3, 4]))

File: 496. Next Greater Element I.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> m;
        stack<int> s;
        for (int num: nums2) {
            while (!s.empty() && s.top() < num) {
                m[s.top()] = num;
                s.pop();
            }
            s.push(num);
        }
        vector<int> ans;
        for (int num: nums1) {
            if (m.find(num) == m.end()) {
                ans.push_back(-1);
            } else {
                ans.push_back(m[num]);
            }
        }
        return ans;
    }
};

int main() {
    vector<int> v1{0,3,5,6};
    vector<int> v2{3,0,6,1,5};
    auto ans = Solution().nextGreaterElement(v1, v2);
    for (int i: ans) cout << i << ' ';
    cout << endl;
    return 0;
}
